core/src/androidMain/kotlin/Peripheral.kt

```diff
@@ -111,7 +112,7 @@ public class AndroidPeripheral internal constructor(
 
     public override suspend fun connect() {
         check(job.isNotCancelled) { "Cannot connect, scope is cancelled for $this" }
-        connectJob.updateAndGet { it ?: createConnectJob() }!!.join()
+        connectJob.updateAndGet { it ?: connectAsync() }!!.await()
```

Comment:
What happens if `await()` is called when the `Deferred` is already awaiting?

Reply:
Multiple callers will all suspend until there is a result. Similar behavior as `join`, it's just expecting a result now. We aren't actually providing a result via `async`, but wanted the behavior of holding onto any failures for us until we've had a chance to call `await`.

Reply:
Does it matter that the `connectJob` is never cleared? Wouldn't disconnecting and then reconnecting just return the result of the first `connectAsync()` without actually doing anything?

Reply:
It's a bit of a dance, so admittedly hard to follow. The dance was necessary as multiple code paths (e.g. `disconnect` invoked, or remote peripheral dropped connection, etc) needed to cause the clearing of the `connectJob`.

On Android, handled via the `Connection`'s `invokeOnClose`:

https://github.com/JuulLabs/kable/blob/fe64627bd19e43f602d7e574c0f148e2977c24df/core/src/androidMain/kotlin/Peripheral.kt#L78-L83

macOS: https://github.com/JuulLabs/kable/blob/fe64627bd19e43f602d7e574c0f148e2977c24df/core/src/macosX64Main/kotlin/Peripheral.kt#L113-L118

JS: https://github.com/JuulLabs/kable/blob/fe64627bd19e43f602d7e574c0f148e2977c24df/core/src/jsMain/kotlin/Peripheral.kt#L231-L237

Reply:
Ah, got it. I was looking for it in the disconnect-related calls. 👍